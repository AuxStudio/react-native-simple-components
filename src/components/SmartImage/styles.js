import { StyleSheet } from 'react-native';

import styleConstants from '../../styleConstants';

const styles = StyleSheet.create({
  container: {},
  backgroundContainer: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: styleConstants.colors.white,
  },
  icon: {
    fontSize: styleConstants.fonts.sizes.icon,
    color: styleConstants.colors.primaryText,
  },
});

export default styles;
