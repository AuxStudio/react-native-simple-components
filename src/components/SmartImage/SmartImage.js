import React from 'react';
import PropTypes from 'prop-types';
import { View, ViewPropTypes, Text, Image, ActivityIndicator } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';

import styles from './styles';
import styleConstants from '../../styleConstants';

export default class SmartImage extends React.Component {
  static propTypes = {
    source: PropTypes.oneOfType([
      PropTypes.shape({
        uri: PropTypes.string,
      }), // remote
      PropTypes.number, // local
    ]),
    iconStyle: Text.propTypes.style,
    style: ViewPropTypes.style,
    loaderColor: PropTypes.string,
  };

  static defaultProps = {
    loaderColor: styleConstants.colors.primaryText,
  };

  state = {
    isLoading: true,
    hasError: false,
    isOffline: false,
  };

  setError = (event) => {
    if (event.nativeEvent.error === 'The Internet connection appears to be offline.') {
      this.setState({
        hasError: true,
        isOffline: true,
      });
    } else {
      this.setState({
        hasError: true,
      });
    }
    this.setLoading(false);
  };

  setLoading = (isLoading) => {
    this.setState({
      isLoading,
    });
  };

  render() {
    let backgroundComponent;

    if (this.state.hasError) {
      const iconName = this.state.isOffline ? 'signal-cellular-off' : 'error-outline';

      backgroundComponent = (
        <View style={styles.backgroundContainer}>
          <Icon name={iconName} style={[styles.icon, this.props.iconStyle]} testID="icon" />
        </View>
      );
    } else if (this.state.isLoading) {
      backgroundComponent = (
        <View style={styles.backgroundContainer}>
          <ActivityIndicator size="large" color={this.props.loaderColor} testID="loader" />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <Image
          source={this.props.source}
          style={this.props.style}
          onLoadEnd={() => this.setLoading(false)}
          onError={this.setError}
        />
        {backgroundComponent}
      </View>
    );
  }
}
