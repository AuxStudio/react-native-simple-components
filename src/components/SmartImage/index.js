import SmartImage from './SmartImage';

export default SmartImage;
