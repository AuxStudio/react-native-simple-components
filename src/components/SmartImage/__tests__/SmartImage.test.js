/* eslint-disable import/first */
import React from 'react';
import renderer from 'react-test-renderer';

import SmartImage from '../';

const IMAGE_SOURCE_LOCAL = 5;

const IMAGE_SOURCE_REMOTE = 'https://d3iw72m71ie81c.cloudfront.net/male-47.jpg';

it('renders a SmartImage', () => {
  expect(
    renderer.create(
      <SmartImage
        source={IMAGE_SOURCE_LOCAL}
        iconStyle={{ color: 'red' }}
        style={{ backgroundColor: 'blue' }}
        loaderColor="green"
      />,
    ),
  ).toMatchSnapshot();
});

it('renders a SmartImage with remote URI', () => {
  expect(renderer.create(<SmartImage source={{ uri: IMAGE_SOURCE_REMOTE }} />)).toMatchSnapshot();
});

it('renders a SmartImage with no props', () => {
  expect(renderer.create(<SmartImage />)).toMatchSnapshot();
});

it('renders a SmartImage and toggles loading', () => {
  const component = renderer.create(<SmartImage source={IMAGE_SOURCE_LOCAL} />);
  const instance = component.getInstance();

  expect(component).toMatchSnapshot();

  // Set offline error
  instance.setLoading(false);
  expect(instance.state.isLoading).toBe(false);
});

it('renders a SmartImage with error', () => {
  const component = renderer.create(<SmartImage source={IMAGE_SOURCE_LOCAL} />);
  const { root } = component;
  const instance = component.getInstance();

  expect(component).toMatchSnapshot();

  // Set offline error
  instance.setError({
    nativeEvent: {
      error: 'Test',
    },
  });
  expect(instance.state.hasError).toBe(true);
  expect(instance.state.isLoading).toBe(false);

  // Should now see icon
  const icon = root.findByProps({ testID: 'icon' });
  expect(icon).toBeDefined();
  expect(icon.props.name).toBe('error-outline');
});

it('renders a SmartImage with offline error', () => {
  const component = renderer.create(<SmartImage source={IMAGE_SOURCE_LOCAL} />);
  const { root } = component;
  const instance = component.getInstance();

  expect(component).toMatchSnapshot();

  // Set offline error
  instance.setError({
    nativeEvent: {
      error: 'The Internet connection appears to be offline.',
    },
  });
  expect(instance.state.hasError).toBe(true);
  expect(instance.state.isOffline).toBe(true);
  expect(instance.state.isLoading).toBe(false);

  // Should now see icon
  const icon = root.findByProps({ testID: 'icon' });
  expect(icon).toBeDefined();
  expect(icon.props.name).toBe('signal-cellular-off');
});
/* eslint-enable */
