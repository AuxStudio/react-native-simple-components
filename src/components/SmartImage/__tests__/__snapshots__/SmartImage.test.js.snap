// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a SmartImage 1`] = `
<View
  style={Object {}}
>
  <Image
    onError={[Function]}
    onLoadEnd={[Function]}
    source={5}
    style={
      Object {
        "backgroundColor": "blue",
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="green"
      hidesWhenStopped={true}
      size="large"
      testID="loader"
    />
  </View>
</View>
`;

exports[`renders a SmartImage with no props 1`] = `
<View
  style={Object {}}
>
  <Image
    onError={[Function]}
    onLoadEnd={[Function]}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#212121"
      hidesWhenStopped={true}
      size="large"
      testID="loader"
    />
  </View>
</View>
`;

exports[`renders a SmartImage with remote URI 1`] = `
<View
  style={Object {}}
>
  <Image
    onError={[Function]}
    onLoadEnd={[Function]}
    source={
      Object {
        "uri": "https://d3iw72m71ie81c.cloudfront.net/male-47.jpg",
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#212121"
      hidesWhenStopped={true}
      size="large"
      testID="loader"
    />
  </View>
</View>
`;
